subsys/bluetooth/audio/bap_unicast_client.c

```diff
@@ -2886,40 +2911,109 @@ static uint8_t unicast_client_ase_read_func(struct bt_conn *conn, uint8_t err,
 		goto fail;
 	}
 
-	unicast_client_ep_set_status(ep, &buf);
+	unicast_client_ep_set_status(ep, buf);
 	unicast_client_ep_subscribe(conn, ep);
 
 	params->func(conn, NULL, ep, params);
 
 	params->num_eps++;
 
-	return BT_GATT_ITER_CONTINUE;
+	err = unicast_client_ase_discover(conn, params, handle);
+	if (err != 0) {
+		LOG_DBG("Failed to read ASE: %d", err);
+
+		params->err = err;
+
+		params->func(conn, NULL, NULL, params);
+	}
+
+	return BT_GATT_ITER_STOP;
 
 fail:
 	params->func(conn, NULL, NULL, params);
 	return BT_GATT_ITER_STOP;
 }
 
-static int unicast_client_ase_discover(struct bt_conn *conn,
-				       struct bt_bap_unicast_client_discover_params *params)
+static uint8_t unicast_client_ase_discover_cb(struct bt_conn *conn,
+					      const struct bt_gatt_attr *attr,
+					      struct bt_gatt_discover_params *discover)
 {
-	LOG_DBG("conn %p params %p", conn, params);
+	struct bt_bap_unicast_client_discover_params *params;
+	struct bt_gatt_chrc *chrc;
+	int err;
+
+	params = CONTAINER_OF(discover, struct bt_bap_unicast_client_discover_params,
+			      discover);
+
+	__ASSERT(params->func != NULL, "params->func was NULL");
+
+	if (attr == NULL) {
+		if (params->num_eps == 0) {
+			LOG_DBG("Unable to find %s ASE",
+				bt_audio_dir_str(params->dir));
+
+			params->err = BT_ATT_ERR_ATTRIBUTE_NOT_FOUND;
+
+			params->func(conn, NULL, NULL, params);
+		} else {
+			/* Else we found all the ASEs */
+			err = unicast_client_ase_cp_discover(conn, params);
+			if (err != 0) {
+				LOG_ERR("Unable to discover ASE Control Point");
+				params->err = BT_ATT_ERR_UNLIKELY;
+
+				params->func(conn, NULL, NULL, params);
+			}
+		}
+
+		return BT_GATT_ITER_STOP;
+	}
+
+	chrc = attr->user_data;
+
+	LOG_DBG("conn %p attr %p handle 0x%04x dir %s",
+		conn, attr, chrc->value_handle, bt_audio_dir_str(params->dir));
+
+	/* Reset to use for long read */
+	reset_read_buf(&uni_cli_insts[bt_conn_index(conn)]);
 
 	params->read.func = unicast_client_ase_read_func;
-	params->read.handle_count = 0u;
+	params->read.handle_count = 1U;
+	params->read.single.handle = chrc->value_handle;
+	params->read.single.offset = 0U;
+
+	err = bt_gatt_read(conn, &params->read);
+	if (err != 0) {
+		LOG_DBG("Failed to read PAC records: %d", err);
+
+		params->err = err;
+
+		params->func(conn, NULL, NULL, params);
+	}
+
+	return BT_GATT_ITER_STOP;
```

Comment:
if there is more to discover, should it not return BT_GATT_ITER_CONTINUE?

Reply:
We could do that, yes. 

So in the case of 3 ASEs, where we want to discover and read them all, we can either do 

Option A):
1) Discover ASE_1
2) Discover ASE_2
3) Discover ASE_3
4) Read ASE_1
5) Read ASE_2
6) Read ASE_3

or 

Option B):
1) Discover ASE_1
2) Read ASE_1
3) Discover ASE_2
4) Read ASE_2
5) Discover ASE_3
6) Read ASE_3

Either way requires 6 GATT requests, so the order does not really matter that much, and we are currently using Option B.

Reply:
If we have similar patterns other places in the host code, we should probably stick with that (pattern).